File: ui/src/components/cityCoords.ts
/**
 * cityToLatLon resolves a (city, country) pair to approximate
 * lat/lon. We keep a small lookup of major cities — enough for typical
 * VPN station names. If the city is unknown we fall back to country
 * centroids, then to null. The renderer treats null as "skip pin".
 *
 * This is intentionally rough; it's a visual indicator on a small map,
 * not a routing decision.
 */

interface LatLon {
  lat: number;
  lon: number;
}

const CITY_TABLE: Record<string, LatLon> = {
  // Russia / CIS
  moscow: { lat: 55.75, lon: 37.62 },
  "saint petersburg": { lat: 59.93, lon: 30.34 },
  "st petersburg": { lat: 59.93, lon: 30.34 },
  petersburg: { lat: 59.93, lon: 30.34 },
  novosibirsk: { lat: 55.04, lon: 82.93 },
  yekaterinburg: { lat: 56.84, lon: 60.6 },
  kazan: { lat: 55.79, lon: 49.12 },
  vladivostok: { lat: 43.12, lon: 131.89 },
  kiev: { lat: 50.45, lon: 30.52 },
  kyiv: { lat: 50.45, lon: 30.52 },
  minsk: { lat: 53.9, lon: 27.57 },
  almaty: { lat: 43.25, lon: 76.92 },
  // Europe
  london: { lat: 51.51, lon: -0.13 },
  paris: { lat: 48.86, lon: 2.35 },
  amsterdam: { lat: 52.37, lon: 4.9 },
  frankfurt: { lat: 50.11, lon: 8.68 },
  berlin: { lat: 52.52, lon: 13.4 },
  munich: { lat: 48.14, lon: 11.58 },
  zurich: { lat: 47.38, lon: 8.54 },
  vienna: { lat: 48.21, lon: 16.37 },
  warsaw: { lat: 52.23, lon: 21.01 },
  prague: { lat: 50.07, lon: 14.43 },
  stockholm: { lat: 59.33, lon: 18.07 },
  helsinki: { lat: 60.17, lon: 24.94 },
  oslo: { lat: 59.91, lon: 10.75 },
  copenhagen: { lat: 55.68, lon: 12.57 },
  dublin: { lat: 53.35, lon: -6.26 },
  madrid: { lat: 40.42, lon: -3.7 },
  barcelona: { lat: 41.39, lon: 2.16 },
  rome: { lat: 41.9, lon: 12.5 },
  milan: { lat: 45.46, lon: 9.19 },
  lisbon: { lat: 38.72, lon: -9.14 },
  athens: { lat: 37.98, lon: 23.73 },
  bucharest: { lat: 44.43, lon: 26.1 },
  istanbul: { lat: 41.01, lon: 28.98 },
  // North America
  "new york": { lat: 40.71, lon: -74.01 },
  nyc: { lat: 40.71, lon: -74.01 },
  "los angeles": { lat: 34.05, lon: -118.24 },
  chicago: { lat: 41.88, lon: -87.63 },
  miami: { lat: 25.76, lon: -80.19 },
  dallas: { lat: 32.78, lon: -96.8 },
  seattle: { lat: 47.61, lon: -122.33 },
  "san francisco": { lat: 37.77, lon: -122.42 },
  toronto: { lat: 43.65, lon: -79.38 },
  montreal: { lat: 45.5, lon: -73.57 },
  vancouver: { lat: 49.28, lon: -123.12 },
  // Asia
  tokyo: { lat: 35.68, lon: 139.69 },
  osaka: { lat: 34.69, lon: 135.5 },
  seoul: { lat: 37.57, lon: 126.98 },
  "hong kong": { lat: 22.32, lon: 114.17 },
  hongkong: { lat: 22.32, lon: 114.17 },
  hk: { lat: 22.32, lon: 114.17 },
  singapore: { lat: 1.35, lon: 103.82 },
  bangkok: { lat: 13.76, lon: 100.5 },
  "kuala lumpur": { lat: 3.14, lon: 101.69 },
  jakarta: { lat: -6.21, lon: 106.85 },
  taipei: { lat: 25.03, lon: 121.57 },
  shanghai: { lat: 31.23, lon: 121.47 },
  beijing: { lat: 39.9, lon: 116.4 },
  shenzhen: { lat: 22.54, lon: 114.06 },
  guangzhou: { lat: 23.13, lon: 113.27 },
  mumbai: { lat: 19.08, lon: 72.88 },
  delhi: { lat: 28.61, lon: 77.21 },
  bangalore: { lat: 12.97, lon: 77.59 },
  dubai: { lat: 25.2, lon: 55.27 },
  "tel aviv": { lat: 32.08, lon: 34.78 },
  // Australia
  sydney: { lat: -33.87, lon: 151.21 },
  melbourne: { lat: -37.81, lon: 144.96 },
  // South America
  "sao paulo": { lat: -23.55, lon: -46.63 },
  "são paulo": { lat: -23.55, lon: -46.63 },
  "buenos aires": { lat: -34.6, lon: -58.38 },
  santiago: { lat: -33.45, lon: -70.66 },
  // Africa
  johannesburg: { lat: -26.2, lon: 28.05 },
  cairo: { lat: 30.04, lon: 31.24 },
  lagos: { lat: 6.52, lon: 3.38 },
};

const COUNTRY_TABLE: Record<string, LatLon> = {
  ru: { lat: 55.75, lon: 37.62 },
  russia: { lat: 55.75, lon: 37.62 },
  ua: { lat: 50.45, lon: 30.52 },
  ukraine: { lat: 50.45, lon: 30.52 },
  by: { lat: 53.9, lon: 27.57 },
  belarus: { lat: 53.9, lon: 27.57 },
  kz: { lat: 43.25, lon: 76.92 },
  kazakhstan: { lat: 43.25, lon: 76.92 },
  us: { lat: 38.9, lon: -77.04 },
  usa: { lat: 38.9, lon: -77.04 },
  "united states": { lat: 38.9, lon: -77.04 },
  ca: { lat: 45.42, lon: -75.7 },
  canada: { lat: 45.42, lon: -75.7 },
  mx: { lat: 19.43, lon: -99.13 },
  mexico: { lat: 19.43, lon: -99.13 },
  uk: { lat: 51.51, lon: -0.13 },
  gb: { lat: 51.51, lon: -0.13 },
  "united kingdom": { lat: 51.51, lon: -0.13 },
  fr: { lat: 48.86, lon: 2.35 },
  france: { lat: 48.86, lon: 2.35 },
  de: { lat: 52.52, lon: 13.4 },
  germany: { lat: 52.52, lon: 13.4 },
  nl: { lat: 52.37, lon: 4.9 },
  netherlands: { lat: 52.37, lon: 4.9 },
  pl: { lat: 52.23, lon: 21.01 },
  poland: { lat: 52.23, lon: 21.01 },
  cz: { lat: 50.07, lon: 14.43 },
  fi: { lat: 60.17, lon: 24.94 },
  finland: { lat: 60.17, lon: 24.94 },
  se: { lat: 59.33, lon: 18.07 },
  sweden: { lat: 59.33, lon: 18.07 },
  no: { lat: 59.91, lon: 10.75 },
  norway: { lat: 59.91, lon: 10.75 },
  dk: { lat: 55.68, lon: 12.57 },
  denmark: { lat: 55.68, lon: 12.57 },
  ie: { lat: 53.35, lon: -6.26 },
  ireland: { lat: 53.35, lon: -6.26 },
  es: { lat: 40.42, lon: -3.7 },
  spain: { lat: 40.42, lon: -3.7 },
  it: { lat: 41.9, lon: 12.5 },
  italy: { lat: 41.9, lon: 12.5 },
  pt: { lat: 38.72, lon: -9.14 },
  ch: { lat: 47.38, lon: 8.54 },
  switzerland: { lat: 47.38, lon: 8.54 },
  at: { lat: 48.21, lon: 16.37 },
  tr: { lat: 41.01, lon: 28.98 },
  jp: { lat: 35.68, lon: 139.69 },
  japan: { lat: 35.68, lon: 139.69 },
  kr: { lat: 37.57, lon: 126.98 },
  cn: { lat: 31.23, lon: 121.47 },
  china: { lat: 31.23, lon: 121.47 },
  hk: { lat: 22.32, lon: 114.17 },
  tw: { lat: 25.03, lon: 121.57 },
  sg: { lat: 1.35, lon: 103.82 },
  singapore: { lat: 1.35, lon: 103.82 },
  th: { lat: 13.76, lon: 100.5 },
  my: { lat: 3.14, lon: 101.69 },
  id: { lat: -6.21, lon: 106.85 },
  in: { lat: 19.08, lon: 72.88 },
  india: { lat: 19.08, lon: 72.88 },
  ae: { lat: 25.2, lon: 55.27 },
  il: { lat: 32.08, lon: 34.78 },
  au: { lat: -33.87, lon: 151.21 },
  australia: { lat: -33.87, lon: 151.21 },
  br: { lat: -23.55, lon: -46.63 },
  ar: { lat: -34.6, lon: -58.38 },
  za: { lat: -26.2, lon: 28.05 },
};

export function cityToLatLon(
  city?: string,
  country?: string,
  _addr?: string,
): LatLon | null {
  const c = (city ?? "").trim().toLowerCase();
  if (c) {
    if (CITY_TABLE[c]) return CITY_TABLE[c];
    // try splitting on common separators ("Tokyo, Japan", "Tokyo - JP")
    for (const sep of [",", "-", "·", "|", "/"]) {
      const idx = c.indexOf(sep);
      if (idx > 0) {
        const head = c.slice(0, idx).trim();
        if (CITY_TABLE[head]) return CITY_TABLE[head];
      }
    }
    // fuzzy: contains a known city name
    for (const key of Object.keys(CITY_TABLE)) {
      if (c.includes(key)) return CITY_TABLE[key];
    }
  }
  const cn = (country ?? "").trim().toLowerCase();
  if (cn && COUNTRY_TABLE[cn]) return COUNTRY_TABLE[cn];
  return null;
}

File: flutter/lib/core/utils/city_coords.dart
/// cityToLatLon — port of the Tauri UI's cityCoords.ts.
/// Resolves a (city, country) pair to approximate lat/lon for map pins.
class LatLon {
  final double lat;
  final double lon;
  const LatLon(this.lat, this.lon);
}

const Map<String, LatLon> _cityTable = {
  // Russia / CIS
  'moscow': LatLon(55.75, 37.62),
  'saint petersburg': LatLon(59.93, 30.34),
  'st petersburg': LatLon(59.93, 30.34),
  'petersburg': LatLon(59.93, 30.34),
  'novosibirsk': LatLon(55.04, 82.93),
  'yekaterinburg': LatLon(56.84, 60.6),
  'kazan': LatLon(55.79, 49.12),
  'vladivostok': LatLon(43.12, 131.89),
  'kiev': LatLon(50.45, 30.52),
  'kyiv': LatLon(50.45, 30.52),
  'minsk': LatLon(53.9, 27.57),
  'almaty': LatLon(43.25, 76.92),
  // Europe
  'london': LatLon(51.51, -0.13),
  'paris': LatLon(48.86, 2.35),
  'amsterdam': LatLon(52.37, 4.9),
  'frankfurt': LatLon(50.11, 8.68),
  'berlin': LatLon(52.52, 13.4),
  'munich': LatLon(48.14, 11.58),
  'zurich': LatLon(47.38, 8.54),
  'vienna': LatLon(48.21, 16.37),
  'warsaw': LatLon(52.23, 21.01),
  'prague': LatLon(50.07, 14.43),
  'stockholm': LatLon(59.33, 18.07),
  'helsinki': LatLon(60.17, 24.94),
  'oslo': LatLon(59.91, 10.75),
  'copenhagen': LatLon(55.68, 12.57),
  'dublin': LatLon(53.35, -6.26),
  'madrid': LatLon(40.42, -3.7),
  'barcelona': LatLon(41.39, 2.16),
  'rome': LatLon(41.9, 12.5),
  'milan': LatLon(45.46, 9.19),
  'lisbon': LatLon(38.72, -9.14),
  'athens': LatLon(37.98, 23.73),
  'bucharest': LatLon(44.43, 26.1),
  'istanbul': LatLon(41.01, 28.98),
  // North America
  'new york': LatLon(40.71, -74.01),
  'nyc': LatLon(40.71, -74.01),
  'los angeles': LatLon(34.05, -118.24),
  'chicago': LatLon(41.88, -87.63),
  'miami': LatLon(25.76, -80.19),
  'dallas': LatLon(32.78, -96.8),
  'seattle': LatLon(47.61, -122.33),
  'san francisco': LatLon(37.77, -122.42),
  'toronto': LatLon(43.65, -79.38),
  'montreal': LatLon(45.5, -73.57),
  'vancouver': LatLon(49.28, -123.12),
  // Asia
  'tokyo': LatLon(35.68, 139.69),
  'osaka': LatLon(34.69, 135.5),
  'seoul': LatLon(37.57, 126.98),
  'hong kong': LatLon(22.32, 114.17),
  'hongkong': LatLon(22.32, 114.17),
  'hk': LatLon(22.32, 114.17),
  'singapore': LatLon(1.35, 103.82),
  'bangkok': LatLon(13.76, 100.5),
  'kuala lumpur': LatLon(3.14, 101.69),
  'jakarta': LatLon(-6.21, 106.85),
  'taipei': LatLon(25.03, 121.57),
  'shanghai': LatLon(31.23, 121.47),
  'beijing': LatLon(39.9, 116.4),
  'shenzhen': LatLon(22.54, 114.06),
  'guangzhou': LatLon(23.13, 113.27),
  'mumbai': LatLon(19.08, 72.88),
  'delhi': LatLon(28.61, 77.21),
  'bangalore': LatLon(12.97, 77.59),
  'dubai': LatLon(25.2, 55.27),
  'tel aviv': LatLon(32.08, 34.78),
  // Australia
  'sydney': LatLon(-33.87, 151.21),
  'melbourne': LatLon(-37.81, 144.96),
  // South America
  'sao paulo': LatLon(-23.55, -46.63),
  'são paulo': LatLon(-23.55, -46.63),
  'buenos aires': LatLon(-34.6, -58.38),
  'santiago': LatLon(-33.45, -70.66),
  // Africa
  'johannesburg': LatLon(-26.2, 28.05),
  'cairo': LatLon(30.04, 31.24),
  'lagos': LatLon(6.52, 3.38),
};

const Map<String, LatLon> _countryTable = {
  'ru': LatLon(55.75, 37.62),
  'russia': LatLon(55.75, 37.62),
  'ua': LatLon(50.45, 30.52),
  'ukraine': LatLon(50.45, 30.52),
  'by': LatLon(53.9, 27.57),
  'belarus': LatLon(53.9, 27.57),
  'kz': LatLon(43.25, 76.92),
  'kazakhstan': LatLon(43.25, 76.92),
  'us': LatLon(38.9, -77.04),
  'usa': LatLon(38.9, -77.04),
  'united states': LatLon(38.9, -77.04),
  'ca': LatLon(45.42, -75.7),
  'canada': LatLon(45.42, -75.7),
  'mx': LatLon(19.43, -99.13),
  'mexico': LatLon(19.43, -99.13),
  'uk': LatLon(51.51, -0.13),
  'gb': LatLon(51.51, -0.13),
  'united kingdom': LatLon(51.51, -0.13),
  'fr': LatLon(48.86, 2.35),
  'france': LatLon(48.86, 2.35),
  'de': LatLon(52.52, 13.4),
  'germany': LatLon(52.52, 13.4),
  'nl': LatLon(52.37, 4.9),
  'netherlands': LatLon(52.37, 4.9),
  'pl': LatLon(52.23, 21.01),
  'poland': LatLon(52.23, 21.01),
  'cz': LatLon(50.07, 14.43),
  'fi': LatLon(60.17, 24.94),
  'finland': LatLon(60.17, 24.94),
  'se': LatLon(59.33, 18.07),
  'sweden': LatLon(59.33, 18.07),
  'no': LatLon(59.91, 10.75),
  'norway': LatLon(59.91, 10.75),
  'dk': LatLon(55.68, 12.57),
  'denmark': LatLon(55.68, 12.57),
  'ie': LatLon(53.35, -6.26),
  'ireland': LatLon(53.35, -6.26),
  'es': LatLon(40.42, -3.7),
  'spain': LatLon(40.42, -3.7),
  'it': LatLon(41.9, 12.5),
  'italy': LatLon(41.9, 12.5),
  'pt': LatLon(38.72, -9.14),
  'ch': LatLon(47.38, 8.54),
  'switzerland': LatLon(47.38, 8.54),
  'at': LatLon(48.21, 16.37),
  'tr': LatLon(41.01, 28.98),
  'jp': LatLon(35.68, 139.69),
  'japan': LatLon(35.68, 139.69),
  'kr': LatLon(37.57, 126.98),
  'cn': LatLon(31.23, 121.47),
  'china': LatLon(31.23, 121.47),
  'tw': LatLon(25.03, 121.57),
  'sg': LatLon(1.35, 103.82),
  'th': LatLon(13.76, 100.5),
  'my': LatLon(3.14, 101.69),
  'id': LatLon(-6.21, 106.85),
  'in': LatLon(19.08, 72.88),
  'india': LatLon(19.08, 72.88),
  'ae': LatLon(25.2, 55.27),
  'il': LatLon(32.08, 34.78),
  'au': LatLon(-33.87, 151.21),
  'australia': LatLon(-33.87, 151.21),
  'br': LatLon(-23.55, -46.63),
  'ar': LatLon(-34.6, -58.38),
  'za': LatLon(-26.2, 28.05),
};

/// Resolves (city, country) to approximate lat/lon, or null if unknown.
LatLon? cityToLatLon({String? city, String? country}) {
  final c = (city ?? '').trim().toLowerCase();
  if (c.isNotEmpty) {
    if (_cityTable.containsKey(c)) return _cityTable[c]!;
    // try splitting on common separators
    for (final sep in [',', '-', '·', '|', '/']) {
      final idx = c.indexOf(sep);
      if (idx > 0) {
        final head = c.substring(0, idx).trim();
        if (_cityTable.containsKey(head)) return _cityTable[head]!;
      }
    }
    // fuzzy: contains a known city name
    for (final key in _cityTable.keys) {
      if (c.contains(key)) return _cityTable[key]!;
    }
  }
  final cn = (country ?? '').trim().toLowerCase();
  if (cn.isNotEmpty && _countryTable.containsKey(cn)) return _countryTable[cn]!;
  return null;
}
